File: Pinball App Android/app/src/main/java/com/pillyliu/pinprofandroid/library/HostedImageCandidateSupport.kt
package com.pillyliu.pinprofandroid.library

private enum class HostedImageCandidatePriority(val rank: Int) {
    PIN_PROF(0),
    OPDB_OR_EXTERNAL(1),
    OTHER(2),
}

internal fun prioritizeHostedImageCandidates(candidates: List<String>): List<String> =
    candidates.sortedWith(
        compareBy<String> { hostedImageCandidatePriority(it).rank }
            .thenBy { it },
    )

internal fun hostedImageLoadTimeoutMs(url: String): Long? =
    when (hostedImageCandidatePriority(url)) {
        HostedImageCandidatePriority.PIN_PROF -> 5_000L
        HostedImageCandidatePriority.OPDB_OR_EXTERNAL -> 6_000L
        HostedImageCandidatePriority.OTHER -> 6_000L
    }

private fun hostedImageCandidatePriority(url: String): HostedImageCandidatePriority {
    val lowercased = url.lowercase()
    return when {
        isPinProfPlayfieldUrl(url) -> HostedImageCandidatePriority.PIN_PROF
        "opdb.org" in lowercased || lowercased.startsWith("http://") || lowercased.startsWith("https://") ->
            HostedImageCandidatePriority.OPDB_OR_EXTERNAL
        else -> HostedImageCandidatePriority.OTHER
    }
}

File: Pinball App 2/Pinball App 2/library/HostedImageCandidateSupport.swift
import Foundation

private enum HostedImageCandidatePriority: Int {
    case pinProf = 0
    case opdbOrExternal = 1
    case other = 2
}

func prioritizeHostedImageCandidates(_ candidates: [URL]) -> [URL] {
    candidates.sorted { lhs, rhs in
        let lhsPriority = hostedImageCandidatePriority(lhs)
        let rhsPriority = hostedImageCandidatePriority(rhs)
        if lhsPriority != rhsPriority {
            return lhsPriority.rawValue < rhsPriority.rawValue
        }
        return lhs.absoluteString < rhs.absoluteString
    }
}

func hostedImageLoadTimeout(for url: URL) -> TimeInterval? {
    switch hostedImageCandidatePriority(url) {
    case .pinProf:
        return 5.0
    case .opdbOrExternal:
        return 6.0
    case .other:
        return 6.0
    }
}

private func hostedImageCandidatePriority(_ url: URL) -> HostedImageCandidatePriority {
    if libraryIsPinProfPlayfieldURL(url) {
        return .pinProf
    }

    let host = url.host?.lowercased() ?? ""
    if host.contains("opdb.org") || !host.isEmpty {
        return .opdbOrExternal
    }

    return .other
}
